### DISCUSSION TITLE
Crashes while loading shared libraries

### DISCUSSION THREAD

**Discussion by pendyalasyam** [android/ndk#1822] [General]
Hi,

This is not related to NDK but I believe people here can help me understand when can the following crashes happen. In our application we are observing some crashes with top of the call stack as below. Has anybody seen such callstack in the past? Or anybody has any idea when can we observe this kind of crash?

I am suspecting deleting library while System.load is loading the library would cause this. Can this be the reason ?

Thanks!!

**CallStack:**
  #00  pc 0x00000000000df854  /apex/com.android.runtime/bin/linker64 (__dl_strcmp+148)
  #01  pc 0x0000000000039a08  /apex/com.android.runtime/bin/linker64 (__dl__Z14find_librariesP19android_namespace_tP6soinfoPKPKcmPS2_PNSt3__16vectorIS2_NS8_9allocatorIS2_EEEEmiPK17android_dlextinfobbPNS9_IS0_NSA_IS0_EEEE+1768)
  #02  pc 0x000000000003c978  /apex/com.android.runtime/bin/linker64 (__dl__Z9do_dlopenPKciPK17android_dlextinfoPKv+2280)
  #03  pc 0x00000000000380e0  /apex/com.android.runtime/bin/linker64 (__loader_android_dlopen_ext+72)
  #04  pc 0x00000000000010cc  /apex/com.android.runtime/lib64/bionic/libdl.so (android_dlopen_ext+12)
  #05  pc 0x0000000000007fc0  /apex/com.android.runtime/lib64/libnativeloader.so (OpenNativeLibraryInNamespace+128)
  #06  pc 0x0000000000007c98  /apex/com.android.runtime/lib64/libnativeloader.so (OpenNativeLibrary+280)
  #07  pc 0x000000000037c7c8  /apex/com.android.runtime/lib64/libart.so (art::JavaVMExt::LoadNativeLibrary(_JNIEnv*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, _jobject*, _jclass*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>*)+2312)
  #08  pc 0x0000000000005184  /apex/com.android.runtime/lib64/libopenjdkjvm.so (JVM_NativeLoad+412)
  #09  pc 0x00000000000bcaf4  /system/framework/arm64/boot.oat (art_jni_trampoline+228)
  #10  pc 0x00000000000c8f90  /system/framework/arm64/boot.oat (java.lang.Runtime.load0+560)
  #11  pc 0x00000000000c9490  /system/framework/arm64/boot.oat (java.lang.System.load+96)

**Comment by pirama-arumuga-nainar** [comment]
@rprichard may be able to help.

**Comment by DanAlbert** [comment]
Are you running with any sanitizers? We've seen crashes in `strcmp` (not `dl_strcmp` though) during initialization while porting tsan, and have started seeing the same thing with some asan configurations. We don't know the cause of the problem yet.

**Reply by pendyalasyam** [reply]
Thanks team for the responses!!

I got local repro of these crashes with some test application.

I have 3 libraries
lib1.so
lib2.so
lib3.so which depends on lib2.so.

From main activity we are loading lib1.so and calling a function in lib1.so to delete lib2.so. lib1.so's function starts a thread to delete the library.
once lib1.so function return to java  code we are trying to load lib2.so.

Meaning while one thread is trying to delete the library, other thread is trying to load the library.

Such cases we are seeing this crash.

**Reply by jfgoog** [reply]
It does not seem surprising to me that your app is crashing, if you delete a library and they try to load it.

Assuming my understanding is correct, what is the question/issue here?
